File: examples/python/08_document_validation.py
"""
Example 08: Document Validation
=================================

Demonstrates validating entire JSON-LD graphs containing multiple
node types against multiple shape definitions.

Use case: Validating an organization's knowledge graph before import.
"""

import json
from jsonld_ex import validate_node, validate_document

# ── 1. Define multiple shapes ────────────────────────────────────

print("=== 1. Multiple Shape Definitions ===\n")

person_shape = {
    "@type": "Person",
    "name": {"@required": True, "@type": "xsd:string", "@minLength": 1},
    "email": {"@required": True, "@pattern": r"^[^@\s]+@[^@\s]+\.[^@\s]+$"},
}

organization_shape = {
    "@type": "Organization",
    "name": {"@required": True, "@type": "xsd:string"},
    "url": {"@required": True, "@type": "xsd:string", "@pattern": r"^https?://"},
    "employeeCount": {"@type": "xsd:integer", "@minimum": 1},
}

product_shape = {
    "@type": "Product",
    "name": {"@required": True, "@type": "xsd:string"},
    "price": {"@type": "xsd:double", "@minimum": 0},
    "sku": {"@required": True, "@type": "xsd:string", "@pattern": r"^[A-Z]{2}-\d{4}$"},
}

shapes = [person_shape, organization_shape, product_shape]
print(f"Defined {len(shapes)} shapes: Person, Organization, Product")

# ── 2. Valid document ────────────────────────────────────────────

print("\n=== 2. Valid Document ===\n")

valid_doc = {
    "@graph": [
        {
            "@id": "#alice",
            "@type": "Person",
            "name": "Alice Johnson",
            "email": "alice@acme.com",
        },
        {
            "@id": "#acme",
            "@type": "Organization",
            "name": "Acme Corp",
            "url": "https://acme.example.com",
            "employeeCount": 500,
        },
        {
            "@id": "#widget",
            "@type": "Product",
            "name": "Super Widget",
            "price": 29.99,
            "sku": "SW-1234",
        },
    ]
}

result = validate_document(valid_doc, shapes)
print(f"Valid: {result.valid}")
print(f"Errors: {len(result.errors)}")

# ── 3. Document with errors across node types ────────────────────

print("\n=== 3. Document with Errors ===\n")

invalid_doc = {
    "@graph": [
        {
            "@id": "#bob",
            "@type": "Person",
            "name": "Bob",
            # Missing required email
        },
        {
            "@id": "#badcorp",
            "@type": "Organization",
            "name": "Bad Corp",
            "url": "not-a-url",        # Doesn't match URL pattern
            "employeeCount": 0,         # Below minimum
        },
        {
            "@id": "#product1",
            "@type": "Product",
            "name": "Gadget",
            "price": -5.00,             # Negative price
            "sku": "invalid-sku",       # Doesn't match SKU pattern
        },
        {
            "@id": "#validperson",
            "@type": "Person",
            "name": "Charlie Davis",
            "email": "charlie@example.com",
            # This one is valid
        },
    ]
}

result = validate_document(invalid_doc, shapes)
print(f"Valid: {result.valid}")
print(f"Total errors: {len(result.errors)}\n")

for err in result.errors:
    print(f"  ✗ {err.path}")
    print(f"    [{err.constraint}] {err.message}\n")

# ── 4. Selective validation (only some types) ────────────────────

print("=== 4. Selective Validation ===\n")

# Only validate Person nodes
result = validate_document(invalid_doc, [person_shape])
print(f"Person-only validation:")
print(f"  Errors: {len(result.errors)}")
for err in result.errors:
    print(f"  ✗ {err.path}: {err.message}")

# ── 5. Mixed valid and invalid with summary ─────────────────────

print("\n=== 5. Validation Summary Report ===\n")

large_doc = {
    "@graph": [
        {"@type": "Person", "@id": f"#p{i}", "name": f"Person {i}", "email": f"p{i}@x.com"}
        for i in range(5)
    ] + [
        {"@type": "Person", "@id": "#bad1", "name": ""},         # Empty name
        {"@type": "Person", "@id": "#bad2", "email": "x@y.com"}, # Missing name
        {"@type": "Person", "@id": "#bad3", "name": "Valid", "email": "invalid"},
    ]
}

result = validate_document(large_doc, [person_shape])

total_nodes = len(large_doc["@graph"])
error_nodes = len(set(e.path.split("/")[0] for e in result.errors))

print(f"  Total nodes:   {total_nodes}")
print(f"  Valid nodes:   {total_nodes - error_nodes}")
print(f"  Invalid nodes: {error_nodes}")
print(f"  Total errors:  {len(result.errors)}")
print(f"  Pass rate:     {(total_nodes - error_nodes) / total_nodes * 100:.0f}%")

if result.errors:
    print(f"\n  Error breakdown:")
    from collections import Counter
    constraint_counts = Counter(e.constraint for e in result.errors)
    for constraint, count in constraint_counts.most_common():
        print(f"    {constraint}: {count}")

File: examples/js/08_document_validation.ts
/**
 * Example 08: Document Validation
 * =================================
 *
 * Demonstrates validating entire JSON-LD graphs containing multiple
 * node types against multiple shape definitions.
 *
 * Use case: Validating an organization's knowledge graph before import.
 *
 * Run: npx ts-node examples/js/08_document_validation.ts
 */

import { validateNode, validateDocument, ShapeDefinition } from '../../packages/js/dist';

// ── 1. Define multiple shapes ────────────────────────────────────

console.log('=== 1. Multiple Shape Definitions ===\n');

const personShape: ShapeDefinition = {
  '@type': 'Person',
  name: { '@required': true, '@type': 'xsd:string', '@minLength': 1 },
  email: { '@required': true, '@pattern': '^[^@\\s]+@[^@\\s]+\\.[^@\\s]+$' },
};

const organizationShape: ShapeDefinition = {
  '@type': 'Organization',
  name: { '@required': true, '@type': 'xsd:string' },
  url: { '@required': true, '@type': 'xsd:string', '@pattern': '^https?://' },
  employeeCount: { '@type': 'xsd:integer', '@minimum': 1 },
};

const productShape: ShapeDefinition = {
  '@type': 'Product',
  name: { '@required': true, '@type': 'xsd:string' },
  price: { '@type': 'xsd:double', '@minimum': 0 },
  sku: { '@required': true, '@type': 'xsd:string', '@pattern': '^[A-Z]{2}-\\d{4}$' },
};

const shapes = [personShape, organizationShape, productShape];
console.log(`Defined ${shapes.length} shapes: Person, Organization, Product`);

// ── 2. Valid document ────────────────────────────────────────────

console.log('\n=== 2. Valid Document ===\n');

const validDoc = {
  '@graph': [
    {
      '@id': '#alice',
      '@type': 'Person',
      name: 'Alice Johnson',
      email: 'alice@acme.com',
    },
    {
      '@id': '#acme',
      '@type': 'Organization',
      name: 'Acme Corp',
      url: 'https://acme.example.com',
      employeeCount: 500,
    },
    {
      '@id': '#widget',
      '@type': 'Product',
      name: 'Super Widget',
      price: 29.99,
      sku: 'SW-1234',
    },
  ],
};

let result = validateDocument(validDoc, shapes);
console.log(`Valid: ${result.valid}`);
console.log(`Errors: ${result.errors.length}`);

// ── 3. Document with errors across node types ────────────────────

console.log('\n=== 3. Document with Errors ===\n');

const invalidDoc = {
  '@graph': [
    {
      '@id': '#bob',
      '@type': 'Person',
      name: 'Bob',
      // Missing required email
    },
    {
      '@id': '#badcorp',
      '@type': 'Organization',
      name: 'Bad Corp',
      url: 'not-a-url',        // Doesn't match URL pattern
      employeeCount: 0,         // Below minimum
    },
    {
      '@id': '#product1',
      '@type': 'Product',
      name: 'Gadget',
      price: -5.00,             // Negative price
      sku: 'invalid-sku',       // Doesn't match SKU pattern
    },
    {
      '@id': '#validperson',
      '@type': 'Person',
      name: 'Charlie Davis',
      email: 'charlie@example.com',
      // This one is valid
    },
  ],
};

result = validateDocument(invalidDoc, shapes);
console.log(`Valid: ${result.valid}`);
console.log(`Total errors: ${result.errors.length}\n`);

for (const err of result.errors) {
  console.log(`  ✗ ${err.path}`);
  console.log(`    [${err.constraint}] ${err.message}\n`);
}

// ── 4. Selective validation (only some types) ────────────────────

console.log('=== 4. Selective Validation ===\n');

// Only validate Person nodes
result = validateDocument(invalidDoc, [personShape]);
console.log('Person-only validation:');
console.log(`  Errors: ${result.errors.length}`);
for (const err of result.errors) {
  console.log(`  ✗ ${err.path}: ${err.message}`);
}

// ── 5. Mixed valid and invalid with summary ─────────────────────

console.log('\n=== 5. Validation Summary Report ===\n');

const largeDoc = {
  '@graph': [
    ...Array.from({ length: 5 }, (_, i) => ({
      '@type': 'Person',
      '@id': `#p${i}`,
      name: `Person ${i}`,
      email: `p${i}@x.com`,
    })),
    { '@type': 'Person', '@id': '#bad1', name: '' },           // Empty name
    { '@type': 'Person', '@id': '#bad2', email: 'x@y.com' },   // Missing name
    { '@type': 'Person', '@id': '#bad3', name: 'Valid', email: 'invalid' },
  ],
};

result = validateDocument(largeDoc, [personShape]);

const totalNodes = largeDoc['@graph'].length;
const errorNodeIds = new Set(result.errors.map((e) => e.path.split('/')[0]));
const errorNodes = errorNodeIds.size;

console.log(`  Total nodes:   ${totalNodes}`);
console.log(`  Valid nodes:   ${totalNodes - errorNodes}`);
console.log(`  Invalid nodes: ${errorNodes}`);
console.log(`  Total errors:  ${result.errors.length}`);
console.log(`  Pass rate:     ${Math.round(((totalNodes - errorNodes) / totalNodes) * 100)}%`);

if (result.errors.length > 0) {
  console.log('\n  Error breakdown:');
  const constraintCounts: Record<string, number> = {};
  for (const err of result.errors) {
    constraintCounts[err.constraint] = (constraintCounts[err.constraint] || 0) + 1;
  }
  const sorted = Object.entries(constraintCounts).sort((a, b) => b[1] - a[1]);
  for (const [constraint, count] of sorted) {
    console.log(`    ${constraint}: ${count}`);
  }
}
